Actions/CheckForUpdates/yamlclass.ps1

```diff
@@ -185,4 +224,255 @@ class Yaml {
+        $result = @()
+        foreach($anchor in $anchors) {
+            $result += $this.GetCustomStepsFromAnchor($job, $anchor.Step, $anchor.Before)
+        }
+        foreach($step in $steps) {
+            if ($step -like 'CustomStep*') {
+                if (-not ($result | Where-Object { $_.Name -eq $step })) {
+                    Write-Host "Custom step '$step' does not belong to a supported anchor"
+                }
+            }
+        }
+        return $result
+    }
+
+    [void] AddCustomStepsToAnchor([string] $job, [hashtable[]] $customSteps, [string] $anchorStep, [bool] $before) {
+        $steps = $this.GetStepsFromJob($job)
+        if (!$steps) {
+            Write-Host "::Warning::Cannot find job '$job'"
+            return
+        }
+        $anchorIdx = $steps.IndexOf($anchorStep)
+        if ($anchorIdx -lt 0) {
+            Write-Host "::Warning::Cannot find anchor step '$anchorStep' in job '$job'"
+            return
+        }
+        foreach($customStep in $customSteps | Where-Object { $_.AnchorStep -eq $anchorStep -and $_.Before -eq $before }) {
+            if ($steps -contains $customStep.Name) {
+                Write-Host "Custom step '$($customStep.Name)' already exists in job '$job'"
+            }
+            else {
+                $anchorStart = 0
+                $anchorCount = 0
+                if ($this.Find("Jobs:/$($job):/steps:/- name: $($anchorStep)", [ref] $anchorStart, [ref] $anchorCount)) {
+                    if ($before) {
+                        $this.Insert($anchorStart-1, @('') + @($customStep.Content | ForEach-Object { "      $_" }))
+                    }
+                    else {
+                        $this.Insert($anchorStart+$anchorCount, @('') + @($customStep.Content | ForEach-Object { "      $_" }))
+                    }
+                }
+            }
+            # Use added step as anchor for next step
+            $anchorStep = $customStep.Name
+        }
+    }
+
+    [void] AddCustomStepsToYaml([string] $job, [hashtable[]] $customSteps, [hashtable[]] $anchors) {
+        foreach($anchor in $anchors) {
+            $this.AddCustomStepsToAnchor($job, $customSteps, $anchor.Step, $anchor.Before)
+        }
+    }
+
+    static [PSCustomObject] GetPermissionsFromArray([string[]] $permissionsArray) {
+        $permissions = [PSCustomObject]@{}
+        $permissionsArray | ForEach-Object {
+            $permissions | Add-Member -MemberType NoteProperty -Name $_.Split(':')[0].Trim() -Value $_.Split(':')[1].Trim()
+        }
+        return $permissions
+    }
+
+    static [string[]] GetPermissionsArray([PSCustomObject] $permissions) {
+        $permissionsArray = @()
+        $permissions.PSObject.Properties.Name | ForEach-Object {
+            $permissionsArray += "$($_): $($permissions."$_")"
+        }
+        return $permissionsArray
+    }
+
+    static [PSCustomObject] MergePermissions([PSCustomObject] $permissions, [PSCustomObject] $permissions2) {

```

Comment:
We do not use read-all or write-all and people cannot customize our workflows to do this.
If they customize permissions, it is because they have steps or jobs requiring additional permissions.